src/fit/src/FitQuadProc.cc

```diff
@@ -0,0 +1,239 @@
+        for (entry[3] = entry[2] + 1; entry[3] < n; entry[3]++) table.push_back(entry);
+
+  return table;
+}
+
+std::array<unsigned int, 4> FitQuadProc::ChoosePMTs(unsigned int nhits) {
+  std::array<unsigned int, 4> pmt_ids;
+  for (int j = 0; j < 4; j++) {
+    while (true) {
+      pmt_ids[j] = CLHEP::RandFlat::shootInt(nhits);
+      bool same = false;
+      for (int k = 0; k < j; k++)
+        if (pmt_ids[j] == pmt_ids[k]) same = true;
+      if (!same) break;
+    }
+  }
+  return pmt_ids;
+}
+
+// Define some helpful inline functions
+
+// Return squared magnitude
+static inline double mag2(const double vec[3]) { return vec[0] * vec[0] + vec[1] * vec[1] + vec[2] * vec[2]; }
+
+// Subtracts the 3-vectors a-b
+static inline void vecsub(double *const ans, const double *const a, const double *const b) {
+  ans[0] = a[0] - b[0];
+  ans[1] = a[1] - b[1];
+  ans[2] = a[2] - b[2];
+}
+
+// Dot product a.b for 3-vectors
+static inline double vecdot(const double *const a, const double *const b) {
+  return a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
+}
+
+// Invert a 3x3 matrix, "m", returning the answer as "ans". This
+// attempts to be highly optimized, minimizing the total number of
+// operations by using the explicit solution and avoiding expensive
+// divisions as much as possible.
+static inline void matinvert(double (*const ans)[3], const double (*const m)[3]) {
+  double denominator =
+      (m[0][2] * (m[1][0] * m[2][1] - m[1][1] * m[2][0]) + m[0][1] * (m[1][2] * m[2][0] - m[1][0] * m[2][2]) +
+       m[0][0] * (m[1][1] * m[2][2] - m[1][2] * m[2][1]));
+
+  if (denominator == 0) {
+    debug << "Quad.cc::vecdot() Matrix cannot be inverted. Check PMTs selected. Need fixing.\n";
+    return;
+  }
+  const double idet = 1. / denominator;
+
+  ans[0][0] = (-m[1][2] * m[2][1] + m[1][1] * m[2][2]) * idet;
+  ans[0][1] = (m[0][2] * m[2][1] - m[0][1] * m[2][2]) * idet;
+  ans[0][2] = (-m[0][2] * m[1][1] + m[0][1] * m[1][2]) * idet;
+  ans[1][0] = (m[1][2] * m[2][0] - m[1][0] * m[2][2]) * idet;
+  ans[1][1] = (-m[0][2] * m[2][0] + m[0][0] * m[2][2]) * idet;
+  ans[1][2] = (m[0][2] * m[1][0] - m[0][0] * m[1][2]) * idet;
+  ans[2][0] = (-m[1][1] * m[2][0] + m[1][0] * m[2][1]) * idet;
+  ans[2][1] = (m[0][1] * m[2][0] - m[0][0] * m[2][1]) * idet;
+  ans[2][2] = (-m[0][1] * m[1][0] + m[0][0] * m[1][1]) * idet;
+}
+
+Processor::Result FitQuadProc::Event(DS::Root *ds, DS::EV *ev) {
+  // FIXME: This should eventually be done automatically
+  // Check if run information needs to be retrieved
+  if (fRun != DS::RunStore::Get()->GetRun(ds)) {
+    fRun = DS::RunStore::Get()->GetRun(ds);
+    BeginOfRun(fRun);
+  }
+
+  std::vector<double> pmtx, pmty, pmtz, pmtt;
+  for (int i : ev->GetAllDigitPMTIDs()) {
+    DS::DigitPMT *pmt = ev->GetOrCreateDigitPMT(i);
+    TVector3 pmtpos = fPMTInfo->GetPosition(pmt->GetID());
+    if (pmt->GetDigitizedTime() > 1e6) {
+      continue;
+    }
+    pmtt.push_back(pmt->GetDigitizedTime());
```

Comment:
We should have a method for how to select which timing to use in the reconstruction algorithms.